diff --git a/install_script.bat b/install_script.bat
index 89390c0..c4c1dc5 100644
--- a/install_script.bat
+++ b/install_script.bat
@@ -25,6 +25,9 @@ REM Check what architecture we are installing on
 if "%PROCESSOR_ARCHITECTURE%"=="AMD64" (
     echo Detected 64 bit system...
     set SYSARCH=64
+) else if "%PROCESSOR_ARCHITECTURE%"=="ARM64" (
+    echo Detected 64 bit system...
+    set SYSARCH=64
 ) else if "%PROCESSOR_ARCHITECTURE%"=="x86" (
     if "%PROCESSOR_ARCHITEW6432%"=="AMD64" (
         echo Detected 64 bit system running 32 bit shell...
diff --git a/yasm.props b/yasm.props
index d68dd3b..84456cf 100644
--- a/yasm.props
+++ b/yasm.props
@@ -15,7 +15,8 @@
       <ObjectFileName>$(IntDir)%(FileName).obj</ObjectFileName>
       <CommandLineTemplate Condition="'$(Platform)' == 'Win32'">"$(YasmPath)"yasm.exe -Xvc -f win32 [AllOptions] [AdditionalOptions] "%(FullPath)"</CommandLineTemplate>
       <CommandLineTemplate Condition="'$(Platform)' == 'x64'">"$(YasmPath)"yasm.exe -Xvc -f win64 [AllOptions] [AdditionalOptions] "%(FullPath)"</CommandLineTemplate>
-      <CommandLineTemplate Condition="'$(Platform)' != 'Win32' and '$(Platform)' != 'x64'">echo YASM not supported on this platform
+	  <CommandLineTemplate Condition="'$(Platform)' == 'ARM64'">"$(YasmPath)"yasm.exe -Xvc -f win64 [AllOptions] [AdditionalOptions] "%(FullPath)"</CommandLineTemplate>
+      <CommandLineTemplate Condition="'$(Platform)' != 'Win32' and '$(Platform)' != 'x64' and '$(Platform)' != 'ARM64'">echo YASM not supported on this platform
 exit 1</CommandLineTemplate>
       <ExecutionDescription>%(Identity)</ExecutionDescription>
     </YASM>
